File: Semester 1/Class10_Computer_Contest_Lv_2_Notes_and_Homework_Nov_13-14/lvl2_hw_sol/lvl2_growing_trees.cpp
#include <bits/stdc++.h>
using namespace std;
int N, h[1002], s[1002], ans = INT_MAX;
int main(){
    scanf("%d", &N);
    for(int i=1; i<=N; i++){
        scanf("%d %d", &h[i], &s[i]);
        for(int j=1; j<i; j++){
            if((h[i] - h[j])*(s[i]-s[j]) > 0) continue;
            int dh = h[j] - h[i], ds = s[i] - s[j];
            if(dh == 0) ans = 0;
            else if(ds!=0 && dh % ds == 0) ans = min(ans, dh/ds);
        }
    }
    if(ans == INT_MAX) printf("-1\n");
    else printf("%d\n", ans);
}

File: Semester 1/Class10_Computer_Contest_Lv_2_Notes_and_Homework_Nov_13-14/lvl2_hw_sol/lvl2_growing_trees.py
import sys
input = sys.stdin.readline
n = int(input())
t = []
for i in range(n):
    t.append(tuple(map(int, input().split())))
ans = float('inf')
for i in range(n):
    for j in range(i+1, n):
        dh, ds = t[i][0] - t[j][0], t[i][1] - t[j][1]
        if dh * ds > 0:
            continue
        if dh == 0:
            ans = 0
        if ds != 0 and dh % ds == 0:
            ans = min(ans, -dh // ds)
print(ans) if ans != float('inf') else print(-1)